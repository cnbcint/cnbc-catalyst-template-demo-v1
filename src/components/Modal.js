import React, { useState, useContext, useEffect } from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

export const ModalContext = React.createContext({
  modals: {},
});

const Container = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  z-index: 200;
  width: 100%;
  background: rgba(0, 0, 0, 0.75);

  height: 100%;
  display: ${(props) => (props.open ? "flex" : "none")};
  padding: 2rem;
  padding-top: 150px;
  box-sizing: border-box;
`;

const Content = styled.div`
  color: white;
`;

const CloseButton = styled.button`
  position: absolute;
  right: 2rem;
  cursor: pointer;
  background: none;
  border: none;
  color: white;
`;

export default function Modal(props) {
  const [open, setOpen] = useState(false);
  const toggle = () => {
    setOpen(!open);
  };
  const modalContext = useContext(ModalContext);
  useEffect(() => {
    modalContext.modals[props.id] = { open, setOpen };
  }, []);
  return (
    <Container open={open} className={props.className ? props.className : ""}>
      <CloseButton
        onClick={toggle}
        className={props.className ? props.className + "__close" : ""}
      >
        close X
      </CloseButton>
      <Content className={props.className ? props.className + "__content" : ""}>
        {props.children}
      </Content>
    </Container>
  );
}

export const ModalController = (props) => {
  const modalContext = useContext(ModalContext);
  return (
    <span
      className={props.className}
      onClick={() => {
        modalContext.modals[props.modalID].setOpen(true);
      }}
    >
      {props.children}
    </span>
  );
};

Modal.propTypes = {
  id: PropTypes.string.isRequired,
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
};
ModalController.defaultProps = {
  className: "",
};
ModalController.propTypes = {
  className: PropTypes.string,
  modalID: PropTypes.string.isRequired,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
};
