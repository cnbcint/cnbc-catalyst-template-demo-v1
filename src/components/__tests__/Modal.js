import React from "react";
import Modal, { ModalController } from "../Modal";
import { render } from "@testing-library/react";

import "jest-styled-components";

const Component = () => (
  <>
    <Modal className="test" id="modal1">
      Testing Content
    </Modal>
    <ModalController modalID="modal1" className="test__open">
      Open Modal 1
    </ModalController>
  </>
);

beforeAll(() => {
  window.IntersectionObserver = class {
    observe() {
      return null;
    }
    disconnect() {}
  };
});

describe("Modal", () => {
  it("Renders correctly", () => {
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Renders the correct content", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName("test__content")[0]
      .textContent;

    expect(target).toMatch("Testing Content");
  });

  it("Renders the open modal button", () => {
    const { container } = render(<Component />);
    const target = container.getElementsByClassName("test__open")[0]
      .textContent;

    expect(target).toMatch("Open Modal 1");
  });

  it("Opens and closes the modal", () => {
    const { container } = render(<Component />);
    const openButton = container.getElementsByClassName("test__open")[0];
    openButton.click();
    const target = container.getElementsByClassName("test")[0];
    expect(target).toHaveStyleRule("display", "flex");

    const closeButton = container.getElementsByClassName("test__close")[0];
    closeButton.click();
    expect(target).toHaveStyleRule("display", "none");
  });

  it("Passes the class name", () => {
    const { container } = render(<Component />);
    expect(container.firstChild.classList.contains("test")).toBe(true);
  });
});

afterAll(() => {
  window.IntersectionObserver = null;
});
