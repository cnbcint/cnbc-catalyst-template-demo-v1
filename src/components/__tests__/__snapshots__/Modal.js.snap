// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Renders correctly 1`] = `
.c0 {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 200;
  width: 100%;
  background: rgba(0,0,0,0.75);
  height: 100%;
  display: none;
  padding: 2rem;
  padding-top: 150px;
  box-sizing: border-box;
}

.c2 {
  color: white;
}

.c1 {
  position: absolute;
  right: 2rem;
  cursor: pointer;
  background: none;
  border: none;
  color: white;
}

<div
  class="c0 test"
>
  <button
    class="c1 test__close"
  >
    close X
  </button>
  <div
    class="c2 test__content"
  >
    Testing Content
  </div>
</div>
`;
